benchmarks/prescribe_dike_injection/prescribed_dike_injection.cc

```diff
@@ -0,0 +1,571 @@
+            prm.declare_entry("Function expression","0.0");
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    PrescribedDikeInjection<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Prescribed dike injection");
+        {
+          AssertThrow( prm.get("Base model") != "Prescribed dike injection",
+                       ExcMessage("You may not use ''prescribed dike injection'' as the base model for itself."));
+
+          // create the base model and initialize its SimulatorAccess base
+          // class; it will get a chance to read its parameters below after we
+          // leave the current section
+          base_model = create_material_model<dim>(prm.get("Base model"));
+          if (SimulatorAccess<dim> *sim = dynamic_cast<SimulatorAccess<dim>*>(base_model.get()))
+            sim->initialize_simulator (this->get_simulator());
+
+          dike_material_injection_fraction = prm.get_double ("Dike material injection fraction");
+          T_bottom_dike = prm.get_double ("Dike bottom temperature");
+          enable_random_dike_generation = prm.get_bool("Enable random dike generation");
+          x_center_dike_generation_zone = prm.get_double ("X center of the dike generation zone");
+          width_dike_generation_zone = prm.get_double ("Width of the dike generation zone");
+          total_refinement_levels = prm.get_double ("Total refinement levels");
+          seed = prm.get_double ("Random number generator seed");
+          ref_top_depth_random_dike = prm.get_double ("Reference top depth of randomly generated dike");
+          range_depth_change_random_dike = prm.get_double ("Range of randomly generated dike depth change");
+          width_random_dike = prm.get_double ("Width of randomly generated dike");
+
+          prm.enter_subsection("Dike injection function");
+          {
+            try
+              {
+                injection_function.parse_parameters(prm);
+              }
+            catch (...)
+              {
+                std::cerr << "ERROR: FunctionParser failed to parse\n"
+                          << "\t Dike injection function\n"
+                          << "with expression \n"
+                          << "\t' " << prm.get("Function expression") << "'";
+                throw;
+              } 
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+
+      // If 'Free Surface' is used, please ensure the 'Surface velocity
+      // projection' is vertical. If the projection is normal, which means
+      // the surface mesh can deform both horizontally and vertically, this
+      // may distort a surface element and give a non-positive volume fraction
+      // in a quadrature point which is invalid.
```

Comment:
Have you run into this particular problem? A non-positive volume fraction would lead to a crash of the run anyway I'd think.

Reply:
Not exactly. Some test models show that it is better to use the vertical projection when using the free surface bc